server/src/main/java/org/elasticsearch/Version.java

```diff
@@ -372,6 +372,8 @@ public static Version fromString(String version) {
     public final byte build;
     public final org.apache.lucene.util.Version luceneVersion;
 
+    private List<Version> declaredVersionsCache;
```

Comment:
I was thinking something like:

```diff
diff --git a/server/src/main/java/org/elasticsearch/Version.java b/server/src/main/java/org/elasticsearch/Version.java
index 1e1993e6d6..48d382ce14 100644
--- a/server/src/main/java/org/elasticsearch/Version.java
+++ b/server/src/main/java/org/elasticsearch/Version.java
@@ -24,6 +24,7 @@ import org.elasticsearch.common.Strings;
 import org.elasticsearch.common.SuppressForbidden;
 import org.elasticsearch.common.io.stream.StreamInput;
 import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.lucene.uid.Versions;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.monitor.jvm.JvmInfo;
 
@@ -372,7 +373,9 @@ public class Version implements Comparable<Version> {
     public final byte build;
     public final org.apache.lucene.util.Version luceneVersion;
 
-    private List<Version> declaredVersionsCache;
+    private static class DeclaredVersionsHolder {
+        static List<Version> declaredVersions = Collections.unmodifiableList(getDeclaredVersions(Version.class));
+    }
 
     Version(int id, org.apache.lucene.util.Version luceneVersion) {
         this.id = id;
@@ -414,17 +417,10 @@ public class Version implements Comparable<Version> {
     public Version minimumCompatibilityVersion() {
         if (major >= 6) {
             // all major versions from 6 onwards are compatible with last minor series of the previous major
-            List<Version> declaredVersions = declaredVersionsCache;
-
-            if (declaredVersions == null) {
-                declaredVersions = Collections.unmodifiableList(getDeclaredVersions(getClass()));
-                declaredVersionsCache = declaredVersions;
-            }
-
             Version bwcVersion = null;
 
-            for (int i = declaredVersions.size() - 1; i >= 0; i--) {
-                final Version candidateVersion = declaredVersions.get(i);
+            for (int i = DeclaredVersionsHolder.declaredVersions.size() - 1; i >= 0; i--) {
+                final Version candidateVersion = DeclaredVersionsHolder.declaredVersions.get(i);
                 if (candidateVersion.major == major - 1 && candidateVersion.isRelease() && after(candidateVersion)) {
                     if (bwcVersion != null && candidateVersion.minor < bwcVersion.minor) {
                         break;
```

Reply:
Looks good to me, I was copying the String#hashCode style. I'll make that change.